src/apps/competitions/unpackers/base_unpacker.py

```diff
@@ -0,0 +1,294 @@
+        {
+            "index": index,
+            "name": phase_name,
+            "description": phase_description,
+            "start": phase_start (datetime.datetime),
+            "end": phase_end (datetime.datetime),
+            # ... See serializer for complete fields list
+            "tasks": [list of indices that should match self.competition['tasks']]
+        }
+        Should call self._validate_phase_ordering() and self._set_phase_statuses()
+        """
+        raise NotImplementedError
+
+    def _unpack_leaderboards(self):
+        """
+        {
+            'id',
+            'primary_index',
+            'title',
+            'key',
+            'columns' : {
+                'id',
+                'computation',
+                'computation_indexes',
+                'title',
+                'key',
+                'sorting',
+                'index',
+            },
+        }
+        """
+        raise NotImplementedError
+
+    def _save_tasks(self):
+        for index, task in self.competition['tasks'].items():
+            if isinstance(task, str):
+                try:
+                    task = Task.objects.get(key=task)
+                except Task.DoesNotExist:
+                    raise CompetitionUnpackingException(f'Task with key {task} does not exist')
+                self.competition['tasks'][index] = task
+            else:
+                for file_type in ['ingestion_program', 'input_data', 'scoring_program', 'reference_data']:
+                    try:
+                        task_file_data = task[file_type]
+                    except KeyError:
+                        # file type not here, moving on
+                        continue
+                    key = self._get_data_key(**task_file_data)
+                    task[file_type] = key
+                serializer = TaskSerializer(
+                    data=task,
+                )
+                serializer.is_valid(raise_exception=True)
+                new_task = serializer.save()
+                self.created_tasks.append(new_task)
+                self.competition['tasks'][index] = new_task
+
+    def _save_solutions(self):
+        for solution in self.competition['solutions']:
+            if 'key' in solution:
+                try:
+                    s = Solution.objects.get(key=solution['key'])
+                except Solution.DoesNotExist:
+                    raise CompetitionUnpackingException(f'Could not find solution with key: {solution["key"]}')
+                for task_index in solution['tasks']:
+                    s.add(self.competition['tasks'][task_index])
+            else:
+                solution['tasks'] = [self.competition['tasks'][index].key for index in solution['tasks']]
+                solution['data'] = self._get_data_key(**solution, file_type='solution')
+                serializer = SolutionSerializer(data=solution)
+                serializer.is_valid(raise_exception=True)
```

Comment:
Ah, do we catch these exceptions properly? This will raise a `ValidationError` I believe, and we catch `CompetitionUnpackErrors`?

Reply:
the save function that calls this:
```python
    def save(self):
        try:
            self._save_tasks()
            self._save_solutions()
            return self._save_competition()
        except Exception as e:
            self._clean()
            raise e
```
so we are catching this exception, and then raising it again?
and then in `unpack_competition` (the function that calls `unpacker.save()`) we do the same thing:
```python
try:
  # ...
  unpacker.save()
  # ...
except Exception as e:
  # ...
  raise e
```

Reply:
comp unpack should only catch `CompUnpackException`